shared/discovery/processor.go

```diff
@@ -0,0 +1,74 @@
+package discovery
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/smartcontractkit/chainlink-ccip/pkg/consts"
+	"github.com/smartcontractkit/chainlink-ccip/pkg/reader"
+	"github.com/smartcontractkit/chainlink-ccip/shared"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-common/pkg/types/ccipocr3"
+)
+
+type Outcome struct {
+}
+
+type Query []byte
+
+type Observation struct {
+	OnRamp map[ccipocr3.ChainSelector][]byte
+}
+
+var _ shared.PluginProcessor[Query, Observation, Outcome] = &ContractDiscoveryProcessor{}
+
+// ContractDiscoveryProcessor is a plugin processor for discovering contracts.
+type ContractDiscoveryProcessor struct {
+	lggr      logger.Logger
+	reader    reader.CCIPReader
+	homechain reader.HomeChain
+	dest      ccipocr3.ChainSelector
+}
+
+func (cdp *ContractDiscoveryProcessor) Query(ctx context.Context, prevOutcome Outcome) (Query, error) {
+	return nil, nil
+}
+
+func (cdp *ContractDiscoveryProcessor) Observation(ctx context.Context, prevOutcome Outcome, query Query) (Observation, error) {
+	contracts, err := cdp.reader.DiscoverContracts(ctx, cdp.dest)
+	if err != nil {
+		if errors.Is(err, consts.ErrContractReaderNotFound) {
+			// not a dest reader, no observations (disable entire processor?).
+			return Observation{}, nil
+		}
+		return Observation{}, fmt.Errorf("unable to discover contracts: %w", err)
+	}
+
+	return Observation{
+		OnRamp: contracts[consts.ContractNameOnRamp],
+	}, nil
+}
+func (cdp *ContractDiscoveryProcessor) ValidateObservation(prevOutcome Outcome, query Query, ao shared.AttributedObservation[Observation]) error {
+	return nil
+}
+func (cdp *ContractDiscoveryProcessor) Outcome(prevOutcome Outcome, query Query, aos []shared.AttributedObservation[Observation]) (Outcome, error) {
+	// come to consensus on the onramp addresses and update the chainreader.
+
+	fChain, err := cdp.homechain.GetFChain()
+	if err != nil {
+		return Outcome{}, fmt.Errorf("unable to get fchain: %w", err)
+	}
+
+	onrampAddrs := make(map[ccipocr3.ChainSelector][][]byte)
+	for _, ao := range aos {
+		for chain, addr := range ao.Observation.OnRamp {
+			onrampAddrs[chain] = append(onrampAddrs[chain], addr)
+		}
+	}
+
+	onrampConsensus := shared.GetConsensusMap(cdp.lggr, "onramp", onrampAddrs, fChain)
+	cdp.reader.Sync(context.Background(), onrampConsensus)
```

Comment:
I know this is still WIP but since outcome is pure any network calls must be avoided (Sync being one of them).

We can still update the ccip reader in `Observation` though, by checking the previous outcome.